### DISCUSSION TITLE
UI themes

### DISCUSSION THREAD

**Discussion by Dliwk** [efroemling/ballistica#243] [Ideas]
- When will the BombSquad codebase start changing the interface, it would probably be cool to give each widget a list of classes (like in HTML): btn_positive, btn_negative, primary, secondary, text_title, text_light, text_bold, etc. After that, we can do sorts of cool things, for example
```python
def on_any_widget_update(widget: ba.Widget):
    current_theme = ba.ui.get_current_theme()
    for widgetclass in widget.classes:
        widget.color = current_theme.color.get(widgetclass)

# Somewhere in theme selector...
def apply_theme(self, theme_name: str) -> None:
    ba.app.config['Current Theme'] = theme_name
    ba.app.config.apply_and_commit()
```

**Comment by efroemling** [comment]
Revamping the UI system is something I've been thinking about for a long time; that code is some of the oldest in the game. As you mention, there should be nice pretty classes for different UI elements instead of the single ba.Widget class and ugly bloated accessor functions that exist now. There's other things I'd love to address too, such as the fact that it's a lot of manual work currently to make keyboard/joystick widget selection work well and feel natural. I'd love for that to happen a bit more automatically. Theming would be another nice thing to support; everything is pretty hard-coded at the moment.

My hope is to write a new better designed base class for widgets sometime in the next few versions, and then we can slowly start replacing the janky old widgets with nicer cleaner new ones.

Are there any other particular features/etc. that you'd like to see in the UI system going forward?

**Reply by Dliwk** [reply]
Well, in general, I would like some kind of layouts (like Qt vertical/horizontal/grid layouts) instead of manually calculating the coordinates, with some parameters like margin and offset. In theory, this will even make it easier to navigate between widgets.

**Comment by MatiasMGz** [comment]
Now that we already have the UI revamp, it's the perfect time to get this implemented!
